rs/nns/cmc/src/main.rs

```diff
@@ -1645,18 +1645,284 @@ async fn fetch_transaction(
+        let occupied_entry = match state.blocks_notified.entry(block_index) {
+            Entry::Occupied(ok) => ok,
+
+            Entry::Vacant(_entry) => {
+                println!(
+                    "[cycles] ERROR: Tried to clear the status of block {}, \
+                     but it already has no status?!",
+                    block_index,
+                );
+                return;
+            }
+        };
+
+        // Make sure the block's status is currently Processing.
+        if &NotificationStatus::Processing != occupied_entry.get() {
+            // Otherwise, do not touch the block's status (and log).
+            println!(
+                "[cycles] ERROR: Tried to clear Processing status of block {} \
+                 but its current status is {:?}",
+                block_index,
+                occupied_entry.get(),
+            );
+            return;
+        }
+
+        occupied_entry.remove();
+    });
+}
+
+/// Ok is returned if the transaction is not eligible for an automatic refund
+/// (because its memo indicates one of the supported operations). This is so
+/// that the caller can use the `?` operator to return early in the case where
+/// automatic refund should be issued.
+///
+/// Otherwise, transaction is eligible for an automatic refund. The rest of
+/// these comments assume that we are in this (interesting) case.
+///
+/// Attempts to transfer the ICP (minus fees) back to the sender (by calling
+/// ledger).
+///
+/// Regardless of whether that ledger call succeeds, Err is returned, but the
+/// value in the Err depends on how the ledger call turns out.
+///
+/// If the ledger call failed, the user can retry.
+///
+/// Like the rest of this canister, uses blocks_notified for "journaling". More
+/// precisely, before calling ledger, there are two things:
+///
+///     1. The block MUST have no status. If it does, this returns Err, and no
+///        ledger call is attempted.
+///
+///     2. The block's status is set to Processing.
+///
+/// If the ledger call succeeds, then the block's status is updated to
+/// AutomaticallyRefunded. Otherwise, if the ledger call fails, then the block's
+/// status is cleared to allow the user to try again. Some reasons the call
+/// might fail:
+///
+///     1. Ledger is unavailable. This could be cause by it being upgraded.
+///
+///     2. Ledger is up, but there is something wrong with our request (e.g.
+///        wrong fee).
+///
+/// It is generally assumed that the arguments are consistent with one another.
+/// E.g. we assume that fetching the block (using incoming_block_index), would
+/// give us the same value as incoming_transaction.
+async fn issue_automatic_refund_if_memo_not_offerred(
+    incoming_block_index: BlockIndex,
+    // This is needed because transaction only has an AccountIdentifier.
+    // Although it is possible to go from PrincipalId + Subaccount to
+    // AccountIdentifier, the reverse is not possible. This is a super confusing
```

Comment:
This commentary "super confusing" is not really useful to have in the code here.